File: libraries/cyclestreets-fragments/src/main/java/net/cyclestreets/ElevationProfileFragment.java
package net.cyclestreets;

import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.v4.app.Fragment;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.LinearLayout;
import android.widget.TextView;

import com.jjoe64.graphview.GraphView;
import com.jjoe64.graphview.GridLabelRenderer;
import com.jjoe64.graphview.LabelFormatter;
import com.jjoe64.graphview.Viewport;
import com.jjoe64.graphview.series.DataPoint;
import com.jjoe64.graphview.series.LineGraphSeries;

import net.cyclestreets.api.DistanceFormatter;
import net.cyclestreets.fragments.R;
import net.cyclestreets.routing.Elevation;
import net.cyclestreets.routing.ElevationFormatter;
import net.cyclestreets.routing.Journey;
import net.cyclestreets.routing.Route;
import net.cyclestreets.routing.Segment;
import net.cyclestreets.routing.Waypoints;

import java.util.ArrayList;
import java.util.List;
import java.util.Locale;

import static net.cyclestreets.util.StringUtils.initCap;

public class ElevationProfileFragment extends Fragment implements Route.Listener {
  private LinearLayout graphHolder;

  @Override
  public View onCreateView(@NonNull final LayoutInflater inflater,
                           final ViewGroup container,
                           final Bundle savedInstanceState) {
    final View elevation = inflater.inflate(R.layout.elevation, container, false);
    graphHolder = elevation.findViewById(R.id.graphview);
    return elevation;
  }

  @Override
  public void onResume() {
    super.onResume();
    Route.onResume();
    Route.registerListener(this);
  }

  @Override
  public void onPause() {
    Route.unregisterListener(this);
    super.onPause();
  }

  @Override
  public void onNewJourney(final Journey journey, final Waypoints waypoints) {
    drawGraph(journey);
    fillInOverview(journey);
  }

  @Override
  public void onResetJourney() {}

  private void drawGraph(final Journey journey) {
    final GraphView graphView = new GraphView(getContext());
    final ElevationFormatter formatter = elevationFormatter();

    List<DataPoint> data = new ArrayList<>();
    for (Elevation elevation : journey.elevation().profile())
      data.add(new DataPoint(elevation.distance(), elevation.elevation()));

    LineGraphSeries<DataPoint> graphSeries = new LineGraphSeries<>(data.toArray(new DataPoint[]{}));
    graphSeries.setDrawBackground(true);

    graphView.addSeries(graphSeries);

    // Allow zooming & scrolling on the x-axis (y-axis remains fixed)
    Viewport viewport = graphView.getViewport();
    viewport.setScalable(true);
    viewport.setYAxisBoundsManual(true);
    viewport.setMinY(formatter.roundHeightBelow(journey.elevation().minimum()));
    viewport.setMaxY(formatter.roundHeightAbove(journey.elevation().maximum()));

    GridLabelRenderer gridLabelRenderer = graphView.getGridLabelRenderer();
    gridLabelRenderer.setGridStyle(GridLabelRenderer.GridStyle.HORIZONTAL);
    gridLabelRenderer.setNumHorizontalLabels(5);
    gridLabelRenderer.setNumVerticalLabels(5);
    gridLabelRenderer.setLabelFormatter(new ElevationLabelFormatter(formatter));
    // we handle y-rounding ourselves, and x-rounding makes labels overlap - see https://github.com/jjoe64/GraphView/issues/413
    gridLabelRenderer.setHumanRounding(false, false);

    graphHolder.removeAllViews();
    graphHolder.addView(graphView);    
  }

  private static class ElevationLabelFormatter implements LabelFormatter {
    private final ElevationFormatter formatter;

    private ElevationLabelFormatter(ElevationFormatter formatter) {
      this.formatter = formatter;
    }

    @Override
    public String formatLabel(double value, boolean isValueX) {
      if (isValueX)
        return (value != 0) ? formatter.distance((int)value) : "";
      return formatter.roundedHeight((int) value);
    }

    @Override
    public void setViewport(Viewport viewport) {}
  }

  public TextView getTextView(int id) {
    return getView().findViewById(id);
  }

  /////////////
  // TODO: The methods below are duplicated in ItineraryFragment.java.  Commonise as default
  //       interface methods once our minSdkVersion is >=24.
  private DistanceFormatter distanceFormatter() {
    return DistanceFormatter.formatter(CycleStreetsPreferences.units());
  }

  private ElevationFormatter elevationFormatter() {
    return ElevationFormatter.formatter(CycleStreetsPreferences.units());
  }

  private void fillInOverview(final Journey journey) {
    Segment.Start start = journey.segments().first();

    setText(R.id.title, journey.name());
    setText(R.id.journeyid, String.format(Locale.getDefault(), "#%,d", journey.itinerary()));
    setText(R.id.routetype, initCap(journey.plan()) + " " + getContext().getString(R.string.elevation_route) + ":");
    setText(R.id.distance, distanceFormatter().total_distance(journey.total_distance()));
    setText(R.id.journeytime, start.totalTime());
    setText(R.id.calories, start.calories());
    setText(R.id.carbondioxide, start.co2());
    setText(R.id.elevation_gain, elevationFormatter().height(journey.elevation().totalElevationGain()));
    setText(R.id.elevation_loss, elevationFormatter().height(journey.elevation().totalElevationLoss()));
  }

  private void setText(int id, String text) {
    getTextView(id).setText(text);
  }
}

File: libraries/cyclestreets-fragments/src/main/java/net/cyclestreets/itinerary/ElevationProfileFragment.kt
package net.cyclestreets.itinerary

import android.graphics.Color
import android.os.Bundle
import androidx.fragment.app.Fragment
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.LinearLayout

import com.jjoe64.graphview.GraphView
import com.jjoe64.graphview.GridLabelRenderer
import com.jjoe64.graphview.LabelFormatter
import com.jjoe64.graphview.Viewport
import com.jjoe64.graphview.series.DataPoint
import com.jjoe64.graphview.series.LineGraphSeries

import net.cyclestreets.fragments.R
import net.cyclestreets.routing.ElevationFormatter
import net.cyclestreets.routing.Journey
import net.cyclestreets.routing.Route
import net.cyclestreets.routing.Waypoints
import net.cyclestreets.util.Theme

import java.util.ArrayList

class ElevationProfileFragment : Fragment(), Route.Listener {
    private lateinit var graphHolder: LinearLayout

    override fun onCreateView(inflater: LayoutInflater,
                              container: ViewGroup?,
                              savedInstanceState: Bundle?): View {
        val elevation = inflater.inflate(R.layout.elevation, container, false)
        graphHolder = elevation.findViewById(R.id.graphview)
        return elevation
    }

    override fun onResume() {
        super.onResume()
        Route.onResume()
        Route.registerListener(this)
    }

    override fun onPause() {
        Route.unregisterListener(this)
        super.onPause()
    }

    override fun onNewJourney(journey: Journey, waypoints: Waypoints) {
        drawGraph(journey)
        fillInOverview(journey, requireView(), requireContext().getString(R.string.elevation_route))
    }

    override fun onResetJourney() {}

    private fun drawGraph(journey: Journey) {
        val graphView = GraphView(context)
        val formatter = elevationFormatter()

        // The elevation data series for the whole route
        val elevationData = ArrayList<DataPoint>()
        for (elevation in journey.elevation.profile())
            elevationData.add(DataPoint(elevation.distance().toDouble(), elevation.elevation().toDouble()))
        val elevationSeries = LineGraphSeries(elevationData.toTypedArray())
        elevationSeries.isDrawBackground = true
        graphView.addSeries(elevationSeries)

        // The elevation data series for the current segment - highlight
        journey.activeSegment()?.let { segment ->
            val segmentEndDistance = segment.cumulativeDistance
            val segmentStartDistance = segmentEndDistance - segment.distance
            val segmentElevationData = elevationData.slice(IntRange(
                elevationData.indexOfFirst { dp -> dp.x >= segmentStartDistance.toDouble() },
                elevationData.indexOfLast { dp -> dp.x <= segmentEndDistance.toDouble() }
            ))
            val segmentElevationSeries = LineGraphSeries(segmentElevationData.toTypedArray())
            segmentElevationSeries.color = Theme.highlightColor(context)
            segmentElevationSeries.backgroundColor = Color.argb(153, 0, 152, 0) //0x99009800
            segmentElevationSeries.isDrawBackground = true
            graphView.addSeries(segmentElevationSeries)
        }

        // Allow zooming & scrolling on the x-axis (y-axis remains fixed)
        val viewport = graphView.viewport
        viewport.isScalable = true
        viewport.isYAxisBoundsManual = true
        viewport.setMinY(formatter.roundHeightBelow(journey.elevation.minimum()))
        viewport.setMaxY(formatter.roundHeightAbove(journey.elevation.maximum()))

        val gridLabelRenderer = graphView.gridLabelRenderer
        gridLabelRenderer.gridStyle = GridLabelRenderer.GridStyle.BOTH
        gridLabelRenderer.numHorizontalLabels = 5
        gridLabelRenderer.numVerticalLabels = 5
        gridLabelRenderer.labelFormatter = ElevationLabelFormatter(formatter)
        // we handle y-rounding ourselves, and x-rounding makes labels overlap - see https://github.com/jjoe64/GraphView/issues/413
        gridLabelRenderer.setHumanRounding(false, false)

        graphHolder.removeAllViews()
        graphHolder.addView(graphView)
    }

    private class ElevationLabelFormatter constructor(private val formatter: ElevationFormatter) : LabelFormatter {
        override fun setViewport(viewport: Viewport) {}

        override fun formatLabel(value: Double, isValueX: Boolean): String {
            return if (isValueX)
                if (value != 0.0) formatter.distance(value.toInt()) else ""
            else formatter.roundedHeight(value.toInt())
        }
    }
}
